File: Data Structures & Algorithms/max-area-of-island/submission-5.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        ROWS = len(grid)
        COLS = len(grid[0])    
        maxArea = 0

        for r in range(ROWS):
            for c in range(COLS):
                if grid[r][c] == 1:
                    area = self.sinkIsland(grid, r, c, ROWS, COLS)
                    maxArea = max(area, maxArea)

        return maxArea


    def sinkIsland(self, grid, r, c, m , n):
        if min(r, c) < 0 or r >= m or c >= n or grid[r][c] == 0:
            return 0

        grid[r][c] = 0
        area = 1
        directions = [[1, 0], [0, 1], [-1, 0], [0, -1]]

        for dr, dc in directions:
            area += self.sinkIsland(grid, r + dr, c + dc, m, n)

        return area

File: Data Structures & Algorithms/islands-and-treasure/submission-5.ts
class Solution {
    /**
     * @param {number[][]} grid
     */
    islandsAndTreasure(grid: number[][]): void {
        const ROWS = grid.length
        const COLS = grid[0].length

        const q: [number, number][] = [] //queue

        for (let r = 0; r < ROWS; r++) {
            for (let c = 0; c < COLS; c++) {
                if (grid[r][c] == 0) {
                    q.push([r, c])
                } 
            }
        }

        const directions = [[1, 0], [0, 1], [-1, 0], [0, -1]]
        const INF = 2147483647
        let distance = 0
        let p = 0

        while (p < q.length) {
            const n = q.length
            for (let i = p; i < n; i++) {
                const [r, c] = q[p]
                p++

                for (const [dr, dc] of directions) {
                    const newRow = dr + r
                    const newCol = dc + c
                    if (newRow >= 0 && newCol >= 0 && newRow < ROWS && newCol < COLS && grid[newRow][newCol] === INF) {
                        grid[newRow][newCol] = distance + 1
                        q.push([newRow, newCol])
                    }
                }
            }

            distance++
        }
    }
}
